sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/ProxyAuthenticationHandler.java

```diff
@@ -0,0 +1,313 @@
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.proxy.ProxyConnectException;
+import io.netty.handler.proxy.ProxyHandler;
+import io.netty.util.AttributeKey;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.azure.core.util.AuthorizationChallengeHandler.PROXY_AUTHENTICATE;
+import static com.azure.core.util.AuthorizationChallengeHandler.PROXY_AUTHENTICATION_INFO;
+import static com.azure.core.util.AuthorizationChallengeHandler.PROXY_AUTHORIZATION;
+
+/**
+ * This class handles authorizing requests being sent through a proxy which require authentication.
+ */
+public final class ProxyAuthenticationHandler extends ProxyHandler {
+    private static final String VALIDATION_ERROR_TEMPLATE = "The '%s' returned in the 'Proxy-Authentication-Info' "
+        + "header doesn't match the value sent in the 'Proxy-Authorization' header. Sent: %s, received: %s.";
+
+    private static final AttributeKey<String> PROXY_AUTHORIZATION_KEY = AttributeKey.newInstance("ProxyAuthorization");
+
+    private static final String NONE = "none";
+    private static final String HTTP = "http";
+
+    private static final String CNONCE = "cnonce";
+    private static final String NC = "nc";
+
+    private static final String AUTH_BASIC = "basic";
+    private static final String AUTH_DIGEST = "digest";
+
+    /*
+     * Proxies use 'CONNECT' as the HTTP method.
+     */
+    private static final String PROXY_METHOD = HttpMethod.CONNECT.name();
+
+    /*
+     * Proxies are always the root path.
+     */
+    private static final String PROXY_URI_PATH = "/";
+
+    /*
+     * Digest authentication to a proxy uses the 'CONNECT' method, these can't have a request body.
+     */
+    private static final Supplier<byte[]> PROXY_ENTITY_BODY = () -> new byte[0];
+
+    private final ClientLogger logger = new ClientLogger(ProxyAuthenticationHandler.class);
+
+    private final AuthorizationChallengeHandler challengeHandler;
+    private final AtomicReference<ChallengeHolder> proxyChallengeHolderReference;
+    private final HttpClientCodec codec;
+    private final AtomicReference<String> authScheme;
+
+    private HttpResponseStatus status;
+
+    public ProxyAuthenticationHandler(InetSocketAddress proxyAddress, AuthorizationChallengeHandler challengeHandler,
+        AtomicReference<ChallengeHolder> proxyChallengeHolderReference) {
```

Comment:
Why is this being passed in as an AtomicReference rather than just a ChallengeHOlder? Are we expecting this to be modified by some other object?

Reply:
Per design multiple instances of `ProxyAuthenticationHandler` will reference this object, this is being done since Netty will close the channel once the request has been completed and it will need to be passed to each channel opened by the client.